src/util/side_metadata/global.rs

```diff
@@ -0,0 +1,581 @@
+use super::helpers::{self, *};
+use crate::util::{constants, memory, Address};
+use std::sync::atomic::{AtomicU16, AtomicU32, AtomicU8, Ordering};
+
+// ** NOTE: **
+//  Regardless of the number of bits in a metadata unit, we always represent its content as a word.
+
+pub(super) const INVALID_SIDEMETADATA_ID: SideMetadataID = SideMetadataID(MAX_METADATA_BITS + 1);
+
+// Starting from zero and increasing by one, this type works as a simple side metadata id
+#[derive(Copy, Clone)]
+pub struct SideMetadataID(usize);
+
+impl SideMetadataID {
+    pub const fn new() -> SideMetadataID {
+        INVALID_SIDEMETADATA_ID
+    }
+
+    pub fn as_usize(&self) -> usize {
+        self.0
+    }
+}
+
+// `align[metadata_id]` is the minimum alignment of the source data for `metadata_id`
+// `meta_bits_num_vec[metadata_id]` stores the number of bits requested for `metadata_id`
+// `meta_base_addr_vec[metadata_id]` stores the starting address of the memory to be mapped for the bits of `metadata_id`
+pub struct SideMetadata {
+    pub(super) align: Vec<usize>,
+    pub(super) meta_bits_num_log_vec: Vec<usize>,
+    pub(super) meta_base_addr_vec: Vec<Address>,
+}
+
+unsafe impl Sync for SideMetadata {}
+
+lazy_static! {
+    pub(super) static ref METADATA_SINGLETON: SideMetadata = SideMetadata {
+        align: Vec::with_capacity(MAX_METADATA_BITS),
+        meta_bits_num_log_vec: Vec::with_capacity(MAX_METADATA_BITS),
+        meta_base_addr_vec: Vec::with_capacity(MAX_METADATA_BITS),
+    };
+}
+
+impl SideMetadata {
+    // FIXME(Javad): check the possibility of a safe implementation.
+    #[allow(clippy::cast_ref_to_mut)]
+    #[allow(clippy::mut_from_ref)]
+    unsafe fn mut_self(&self) -> &mut Self {
+        &mut *(self as *const _ as *mut _)
+    }
+
+    /// Requests a number of metadata bits and returns a `SideMetadataID`.
+    /// This ID is used for the future references to these bits and allows choosing between bit sets (e.g. global and policy-specific bits, each have an ID).
+    ///
+    /// # Arguments
+    ///
+    /// * `number_of_bits` - The number of bits per source data unit (e.g. per object).
+    ///     Currently, the maximum metadata size per data unit is a word (usize).
+    ///
+    /// * `align` - The minimum alignment of the source data.
+    ///     The minimum data granularity is a word, which means the minimum value of this argument is 2 in 32-bits, and 3 in 64 bits systems.
+    ///
+    pub fn request_meta_bits(number_of_bits: usize, align: usize) -> SideMetadataID {
```

Comment:
`log_bytes_align` makes more sense, I'll change it.

Good point on the function's safety.
My opinion is that we protect this function with a lock and don't expose the unsafety to the user, because I don't think there is anything inherently unsafe in this function, from the point of view of a user.

Let me know if you have a different opinion

Reply:
That sounds good.